### PR TITLE
cnf network: add-metallb-frr-test-cases

### PR DESCRIPTION
**Pull request by gkopels**
Last two test cases of frrk8 epic

### CODE REVIEW

**Comment by fedepaol** [comment]
did a rough check on what the tests are covering, lgtm from that side. I didn't go deeply in what the code does / how it's written

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 1]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
+			vlanID, err := NetConfig.GetVLAN()
+			Expect(err).ToNot(HaveOccurred(), "Fail to set vlanID")
+
+			By("Removing secondary interface on worker-0")
+			secIntWorker0Policy := nmstate.NewPolicyBuilder(APIClient, "sec-int-worker0", NetConfig.WorkerLabelMap).
+				WithAbsentInterface(fmt.Sprintf("%s.%d", srIovInterfacesUnderTest[0], vlanID))
+			err = netnmstate.UpdatePolicyAndWaitUntilItsAvailable(netparam.DefaultTimeout, secIntWorker0Policy)
+			Expect(err).ToNot(HaveOccurred(), "Failed to update NMState network policy")
+
+			By("Removing secondary interface on worker-1")
```
Comment (evgenLevin) [inline tests/cnf/core/network/metallb/tests/frrk8-tests.go:432]: please replace worker-1

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 2]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
 
-			By("Clean metallb operator and test namespaces")
+			srIovInterfacesUnderTest, err := NetConfig.GetSriovInterfaces(1)
+			Expect(err).ToNot(HaveOccurred(), "Failed to retrieve SR-IOV interfaces for testing")
+
+			vlanID, err := NetConfig.GetVLAN()
+			Expect(err).ToNot(HaveOccurred(), "Fail to set vlanID")
+
+			By("Removing secondary interface on worker-0")
+			secIntWorker0Policy := nmstate.NewPolicyBuilder(APIClient, "sec-int-worker0", NetConfig.WorkerLabelMap).
```
Comment (evgenLevin) [inline tests/cnf/core/network/metallb/tests/frrk8-tests.go:427]: please replace worker-0

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 3]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
+			err = netnmstate.UpdatePolicyAndWaitUntilItsAvailable(netparam.DefaultTimeout, secIntWorker0Policy)
+			Expect(err).ToNot(HaveOccurred(), "Failed to update NMState network policy")
+
+			By("Removing secondary interface on worker-1")
+			secIntWorker1Policy := nmstate.NewPolicyBuilder(APIClient, "sec-int-worker1", NetConfig.WorkerLabelMap).
+				WithAbsentInterface(fmt.Sprintf("%s.%d", srIovInterfacesUnderTest[0], vlanID))
+			err = netnmstate.UpdatePolicyAndWaitUntilItsAvailable(netparam.DefaultTimeout, secIntWorker1Policy)
+			Expect(err).ToNot(HaveOccurred(), "Failed to update NMState network policy")
+
+			By("Removing secondary interfaces")
```
Comment (evgenLevin) [inline tests/cnf/core/network/metallb/tests/frrk8-tests.go:438]: wrong BY

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 4]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
-				LabelSelector: "app=frr-k8s",
+				By("Waiting until the new frr-k8s-webhook-server deployment is in Ready state.")
+				frrk8sWebhookDeployment, err := deployment.Pull(
+					APIClient, frrK8WebHookServer, NetConfig.MlbOperatorNamespace)
+				Expect(err).ToNot(HaveOccurred(), "Fail to pull frr-k8s-webhook-server")
+				Expect(frrk8sWebhookDeployment.IsReady(30*time.Second)).To(BeTrue(),
+					"frr-k8s-webhook-server deployment is not ready")
+
+				By("Creating BGP Peers")
+				createBGPPeerAndVerifyIfItsReady(ipv4metalLbIPList[0], "", 64500,
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:335]: replaced everywhere found for 64500 and 64501

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 5]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
+					false, 0, frrk8sPods)
+
+				By("Create first frrconfiguration that receieves a single route")
+				createFrrConfiguration(frrConfigFiltered1, ipv4metalLbIPList[0],
+					64500, []string{externalAdvertisedIPv4Routes[0], externalAdvertisedIPv6Routes[0]},
+					false, false)
+
+				By("Verify Worker-0 Node state update")
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:343]: replaced everywhere found for Worker-0 and Worker-1

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 6]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
+
+					// Return the routes to be checked
+				}, 60*time.Second, 5*time.Second).Should(SatisfyAll(
+					ContainSubstring(fmt.Sprintf("permit %s", externalAdvertisedIPv4Routes[0])),
+					Not(ContainSubstring(fmt.Sprintf("permit %s", externalAdvertisedIPv4Routes[1]))),
+				), "Fail to find all expected received routes")
+
+				By("Create second frrconfiguration that receives a single route")
+				createFrrConfiguration(frrConfigFiltered2, ipv4metalLbIPList[0],
+					64500, []string{externalAdvertisedIPv4Routes[1], externalAdvertisedIPv6Routes[1]},
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:361]: done

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 7]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
 			}
 
-			By("Clean metallb operator and test namespaces")
+			srIovInterfacesUnderTest, err := NetConfig.GetSriovInterfaces(1)
+			Expect(err).ToNot(HaveOccurred(), "Failed to retrieve SR-IOV interfaces for testing")
+
+			vlanID, err := NetConfig.GetVLAN()
+			Expect(err).ToNot(HaveOccurred(), "Fail to set vlanID")
+
+			By("Removing secondary interface on worker-0")
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:426]: done

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go (hunk 1 above)
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:432]: done

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go (hunk 2 above)
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:427]: done

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 8]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
+				createFrrConfiguration(frrConfigFiltered1, ipv4metalLbIPList[0],
+					64500, []string{externalAdvertisedIPv4Routes[0], externalAdvertisedIPv6Routes[0]},
+					false, false)
+
+				By("Verify Worker-0 Node state update")
+				Eventually(func() string {
+					// Get the routes
+					frrNodeState, err := metallb.ListFrrNodeState(APIClient)
+					Expect(err).ToNot(HaveOccurred(), "Failed to verify BGP routes")
+					fmt.Println("frrNodeState[0].Objects.Status.RunningConfig",
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:348]: Removed

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 9]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
+				Expect(frrk8sWebhookDeployment.IsReady(30*time.Second)).To(BeTrue(),
+					"frr-k8s-webhook-server deployment is not ready")
+
+				By("Creating BGP Peers")
+				createBGPPeerAndVerifyIfItsReady(ipv4metalLbIPList[0], "", 64500,
+					false, 0, frrk8sPods)
+
+				By("Create first frrconfiguration that receieves a single route")
+				createFrrConfiguration(frrConfigFiltered1, ipv4metalLbIPList[0],
+					64500, []string{externalAdvertisedIPv4Routes[0], externalAdvertisedIPv6Routes[0]},
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:340]: replaced every found

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:361]: done

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go (hunk 3 above)
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:438]: Changed

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 10]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
+			}
+
+			By("Reset metallb operator namespaces")
 			resetOperatorAndTestNS()
 		})
 
 		It("Validate a FRR node receives and sends IPv4 and IPv6 routes from an IBGP multihop FRR instance",
 			reportxml.ID("74278"), func() {
 
+				By("Creating a new instance of MetalLB Speakers on workers")
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:454]: Placed them in beforeEach.

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 11]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
 				verifyMetalLbBGPSessionsAreUPOnFrrPod(frrPod, removePrefixFromIPList(ipv4NodeAddrList))
 
 				By("Validating BGP route prefix")
-				validatePrefix(frrPod, "IPv4", removePrefixFromIPList(nodeAddrList), addressPool, 32)
+				validatePrefix(frrPod, "IPv4", removePrefixFromIPList(nodeAddrList),
+					addressPool, 32)
 
 				By("Create a frrconfiguration allow all for EBGP multihop")
-				createFrrConfiguration("frrconfig-allow-all", "172.16.0.1", 64501,
+				createFrrConfiguration(frrCongigAllowAll, frrExternalMasterIPAddress, 64501,
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:622]: done

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 12]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
+				By("Collecting information before test")
+				frrk8sPods, err := pod.List(APIClient, NetConfig.MlbOperatorNamespace, metav1.ListOptions{
+					LabelSelector: frrNodeLabel,
+				})
+				Expect(err).ToNot(HaveOccurred(), "Failed to list speaker pods")
+
+				By("Setting test iteration parameters")
+				_, _, _, _, addressPool, _, err :=
+					metallbenv.DefineIterationParams(
+						ipv4metalLbIPList, ipv6metalLbIPList, ipv4NodeAddrList, ipv6NodeAddrList, "IPv4")
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:662]: Replaced everywhere found

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 13]
```diff
@@ -748,3 +969,55 @@ func buildRoutesMapWithSpecificRoutes(podList []*pod.Builder, nextHopList []stri
...
+	secondaryInterface := nmstate.NewPolicyBuilder(APIClient, policyName, map[string]string{
+		"kubernetes.io/hostname": nodeName,
+	})
+
+	secondaryInterface.WithVlanInterfaceIP(interfaceName, ipv4Address, ipv6Address, vlanID)
+
+	_, err := secondaryInterface.Create()
+
+	if err != nil {
+		return fmt.Errorf("fail to create secondary interface: %s.+%d", interfaceName, vlanID)
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:984]: replaced

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 14]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
+					tsparams.HubMacVlanNADName,
+					[]string{fmt.Sprintf("%s/24", hubSecIntIPv4Addresses[1])},
+					[]string{fmt.Sprintf("%s/24", hubIPv4ExternalAddresses[1])})
+
+				By("Creating FRR Hub Worker-0 Pod")
+				_ = createFrrHubPod(hubPodWorker0,
+					workerNodeList[0].Object.Name, createHubConfigMapSecInt.Definition.Name, []string{},
+					hub0BRStaticSecIntIPAnnotation)
+
+				By("Creating FRR Hub Worker-1 Pod")
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:725]: removed

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:426]: replaced

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go (hunk 1 above)
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:432]: replaced

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go (hunk 2 above)
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:427]: replaced

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 15]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
+			Expect(err).ToNot(HaveOccurred(), "Fail to set vlanID")
+
+			By("Removing secondary interface on worker-0")
+			secIntWorker0Policy := nmstate.NewPolicyBuilder(APIClient, "sec-int-worker0", NetConfig.WorkerLabelMap).
+				WithAbsentInterface(fmt.Sprintf("%s.%d", srIovInterfacesUnderTest[0], vlanID))
+			err = netnmstate.UpdatePolicyAndWaitUntilItsAvailable(netparam.DefaultTimeout, secIntWorker0Policy)
+			Expect(err).ToNot(HaveOccurred(), "Failed to update NMState network policy")
+
+			By("Removing secondary interface on worker-1")
+			secIntWorker1Policy := nmstate.NewPolicyBuilder(APIClient, "sec-int-worker1", NetConfig.WorkerLabelMap).
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:433]: replaced

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 16]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
+					[]string{fmt.Sprintf("%s/24", hubIPv4ExternalAddresses[1])})
+
+				By("Creating MetalLb Hub pod configMap")
+				hubConfigMap := createHubConfigMap("frr-hub-node-config")
+
+				By("Creating FRR Hub Worker-0 Pod")
+				_ = createFrrHubPod(hubPodWorker0,
+					workerNodeList[0].Object.Name, hubConfigMap.Definition.Name, []string{}, hub0BRstaticIPAnnotation)
+
+				By("Creating FRR Hub Worker-1 Pod")
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:509]: replaced

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 17]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
+				By("Setting test iteration parameters")
+				_, _, _, _, addressPool, _, err :=
+					metallbenv.DefineIterationParams(
+						ipv4metalLbIPList, ipv6metalLbIPList, ipv4NodeAddrList, ipv6NodeAddrList, "IPv4")
+				Expect(err).ToNot(HaveOccurred(), "Fail to set iteration parameters")
+
+				By("create a secondary IP address on the worker node 0")
+				err = createSecondaryInterfaceOnNode("sec-int-worker0", workerNodeList[0].Definition.Name,
+					srIovInterfacesUnderTest[0], frrNodeSecIntIPv4Addresses[0], "2001:100::254", vlanID)
+				Expect(err).ToNot(HaveOccurred(), "Fail to create a secondary interface on worker0")
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:668]: replaced

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 18]
```diff
@@ -328,207 +301,454 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
+
+				By("create a secondary IP address on the worker node 1")
+				err = createSecondaryInterfaceOnNode("sec-int-worker1", workerNodeList[1].Definition.Name,
+					srIovInterfacesUnderTest[0], frrNodeSecIntIPv4Addresses[1], "2001:100::253", vlanID)
+				Expect(err).ToNot(HaveOccurred(), "Fail to create a secondary interface on worker1")
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:673]: replaced

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 19]
```diff
@@ -96,49 +110,40 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
-					Expect(err).ToNot(HaveOccurred(), "Failed to verify BGP routes")
-
-					// Return the routes to be checked
-					return routes
-				}, 60*time.Second, 5*time.Second).Should(SatisfyAll(
-					Not(ContainSubstring(prefixToBlock)),              // First IP address
-					ContainSubstring(externalAdvertisedIPv4Routes[1]), // Second IP address
-				), "Fail to find all expected received routes")
+				verifyReceivedRoutes(frrk8sPods, externalAdvertisedIPv4Routes[1])
+				verifyBlockedRoutes(frrk8sPods, prefixToBlock)
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go:147]: added

File: tests/cnf/core/network/metallb/tests/frrk8-tests.go [hunk 20]
```diff
@@ -328,207 +302,402 @@ var _ = Describe("FRR", Ordered, Label(tsparams.LabelFRRTestCases), ContinueOnFa
...
+				createSecondaryInterfaceOnNode("sec-int-worker0", workerNodeList[0].Definition.Name,
+					srIovInterfacesUnderTest[0], frrNodeSecIntIPv4Addresses[0], "2001:100::254", vlanID)
+
+				By("create a secondary IP address on the worker node 1")
+				createSecondaryInterfaceOnNode("sec-int-worker1", workerNodeList[1].Definition.Name,
+					srIovInterfacesUnderTest[0], frrNodeSecIntIPv4Addresses[1], "2001:100::253", vlanID)
+
+				By("Adding static routes to the speakers")
+				speakerRoutesMap := buildRoutesMapWithSpecificRoutes(frrk8sPods, hubSecIntIPv4Addresses)
+				Expect(err).ToNot(HaveOccurred(), "Failed to build speaker route map")
```
[the remark this answers is not in the document]
Comment (gkopels) [reply tests/cnf/core/network/metallb/tests/frrk8-tests.go]: removed

**Review by evgenLevin — APPROVED** [review]
/lgtm

**Review by ajaggapa — APPROVED** [review]
lgtm

**Pull request merged**